Use more fastutil collections

jtraglia opened the pull request:
This is an extension of #4984. It converts the following collections:

* `Set<Integer>` -> `IntSet`
* `List<Integer>` -> `IntList`
* `Collection<Integer>` -> `IntCollection`
* `Iterable<Integer>` -> `IntIterable`

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

jtraglia committed:
Convert a bunch of collections to fastutil

jtraglia committed:
Use IntList instead of IntArrayList for new lists

jtraglia committed:
Convert some more types

I'm still not quite finished.

jtraglia committed:
Maybe fix assemble check problem

jtraglia committed:
Convert a few more types

ajsutton commented on beacon/validator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/SyncCommitteePerformanceTracker.java:
```diff
@@ -201,7 +202,7 @@ private synchronized int countIncludedMessages(
         continue;
       }
-      for (Integer committeeIndex : committeeIndices) {
+      for (int committeeIndex : committeeIndices) {
```
I don't think this is taking advantage of the `IntSet` - it will create a `Iterator<Integer>` which wraps each value and then the value is unwrapped back to `int`.  We'd need to use `IntIterator` directly and call `nextInt()` rather than just `next()`. In this case it may be better to use an `IntStream` instead like:
```
numberOfIncludedMessages += committeeIndices.intStream().filter(syncCommitteeBits::getBit).count();
```

jtraglia replied:
Ah I didn't think of that. I think you're right. I'll change it to one of those.

ajsutton commented on data/serializer/src/main/java/tech/pegasys/teku/api/response/v1/validator/SyncCommitteeDuty.java:
```diff
@@ -55,7 +56,7 @@
       @JsonProperty("pubkey") final BLSPubKey pubkey,
       @JsonProperty("validator_index") final UInt64 validatorIndex,
-      @JsonProperty("validator_sync_committee_indices") final Set<Integer> committeeIndices) {
+      @JsonProperty("validator_sync_committee_indices") final IntSet committeeIndices) {
```
Basically anything with `@JsonProperty` will need to stay declared using the `Integer` form to ensure our OpenAPI docs come out right.  We *might* be able to get away with actually passing it an `IntSet` as the implementation but I'm not sure about that - we'll have to test it.

jtraglia replied:
Good to know. I'll revert those instances.

ajsutton reviewed (commented):
This has wound up being a really big PR.  It's probably worth doing this in smaller steps to make reviewing simpler and make sure we pick up on all the details like using `IntIterator` instead of a `Iterator<Integer>` and other gotchas.  I know a bunch of things wind up inter-related but if we can find more contained areas to do at each step that would be good.  For example the performance tracker classes could be converted over in one PR and `ValidatorApiHandler` in another.

jtraglia committed:
Revert @JsonProperty type change

jtraglia committed:
Convert for-each loops to use primitives

ajsutton commented on infrastructure/ssz/src/main/java/tech/pegasys/teku/infrastructure/ssz/collections/impl/SszBitvectorImpl.java:
```diff
@@ -74,8 +74,8 @@ public SszBitvector rightShift(int n) {
-  public List<Integer> getAllSetBits() {
-    return value.streamAllSetBits().boxed().collect(Collectors.toList());
+  public IntList getAllSetBits() {
+    return IntArrayList.toList(value.streamAllSetBits().boxed().mapToInt(Integer::intValue));
```
This is boxing and then unboxing the ints unnecessarily.  Should just be:
```suggestion
    return IntArrayList.toList(value.streamAllSetBits());
```

jtraglia replied:
Nice spot. That was a bit redundant. Will fix.

ajsutton commented on infrastructure/ssz/src/test/java/tech/pegasys/teku/infrastructure/ssz/collections/SszBitlistTest.java:
```diff
@@ -99,7 +99,7 @@ void testSszRoundtrip(SszBitlist bitlist1) {
     SszBitlist bitlist2 = bitlist1.getSchema().sszDeserialize(ssz1);
 
-    assertThat(bitlist2.getAllSetBits()).isEqualTo(bitlist1.getAllSetBits());
+    assertThat(bitlist2.getAllSetBits().equals(bitlist1.getAllSetBits())).isTrue();
```
hmmm, we don't want to lose the nice error messages we get from `isEqualTo`. Simplest answer is to just use:
```
    assertThat((List<Integer>) bitlist2.getAllSetBits()).isEqualTo(bitlist1.getAllSetBits());
```
but it's not particularly pretty.

I've pushed a commit that adds a new testFixtures source tree to `:infrastructure:collections` with a `PrimitiveCollectionsAssert` to hide the ugliness a bit.  It does mean tests will wind up boxing the ints when doing assertions but that's fine and gives us all the nice assertions and errors messages that assertj provides.

ajsutton commented on validator/client/src/main/java/tech/pegasys/teku/validator/client/ValidatorIndexProvider.java:
```diff
@@ -101,13 +102,14 @@ private void logNewValidatorIndices(final Map<BLSPublicKey, Integer> knownValida
-                .flatMap(validator -> getValidatorIndex(validator.getPublicKey()).stream())
-                .collect(toList()));
+            IntArrayList.toList(
+                ownedValidators.getActiveValidators().stream()
+                    .flatMap(validator -> getValidatorIndex(validator.getPublicKey()).stream())
+                    .mapToInt(Integer::intValue)));
```
As a follow up we could change the `Map<BLSPublicKey, Integer> validatorIndexesByPublicKey` to `Object2IntMap<BLSPublicKey>` so the values can be primitive ints.  Not a high priority I don't think and definitely not worth adding to this PR.

jtraglia replied:
Good idea. I'll keep this in mind afterwards.

ajsutton commented on validator/remote/src/main/java/tech/pegasys/teku/validator/remote/RemoteValidatorApiHandler.java:
```diff
@@ -187,9 +189,10 @@ private SyncCommitteeDuties responseToSyncCommitteeDuties(final PostSyncDutiesRe
+                        IntOpenHashSet.toSet(
+                            duty.committeeIndices.stream()
+                                .map(UInt64::intValue)
+                                .mapToInt(Integer::intValue))))
```
Could just use:
```suggestion
                                .mapToInt(UInt64::intValue))))
```

ajsutton replied:
Actually I think there are probably quite a few places we could do this - worth searching for `mapToInt(Integer::intValue)` to find them I think - I only just realised it when I got to the end of reviewing. :)

jtraglia replied:
Oh yeah. That looks way better. Will update all instances.

ajsutton committed:
Introduce PrimitiveCollectionAssert

ajsutton reviewed (commented):
Looking great. Getting to the final details here.

There is a pretty long stream of work we can likely do to continue this switch which is great, but don't feel you have to keep on this particular issue. It's the kind of thing we can just keep picking off bit by bit, particularly focussing on areas that are actually having a performance impact.

jtraglia committed:
Clean up some integer conversions

There were some `.map(UInt64::intValue).mapToInt(Integer::intValue)` that could
be converted to `.mapToInt(UInt64::intValue)`. Also removed some maps and used
`.intStream()` instead of `.stream()`.

jtraglia committed:
Merge branch 'more-fastutil' of github.com:jtraglia/teku into more-fastutil

ajsutton reviewed (approved):
LGTM. Thanks for the amount of work you're putting into this.

ajsutton committed:
Merge branch 'master' into more-fastutil

ajsutton committed:
Merge branch 'master' into more-fastutil
